accounts-db/src/tiered_storage/hot.rs

```diff
@@ -457,25 +493,142 @@ impl HotStorageWriter {
             storage: TieredStorageFile::new_writable(file_path)?,
         })
     }
+
+    /// Persists an account with the specified information and returns
+    /// the stored size of the account.
+    fn write_account(
+        &self,
+        lamports: u64,
+        rent_epoch: Epoch,
+        account_data: &[u8],
+        executable: bool,
+        account_hash: &AccountHash,
+    ) -> TieredStorageResult<usize> {
+        let optional_fields = AccountMetaOptionalFields {
+            rent_epoch: (rent_epoch != Epoch::MAX).then_some(rent_epoch),
+            account_hash: (*account_hash != AccountHash(Hash::default())).then_some(*account_hash),
+        };
```

Comment:
Optional fields are Option<T>.  The reasons why we have a default for None are:

* In `rent_epoch()` API under ReadableAccount and `hash()` API under StoredAccountMeta, they both require us to provide a value currently.  So even in the None case, we need to return a value there.  If we want to change it to Option, they are the right API to change.

```
    pub fn hash(&self) -> &'storage AccountHash {
```

```
    fn rent_epoch(&self) -> Epoch {
```

* `Epoch::MAX` means the account is rent-exempt, and the majority of the accounts satisfy this.  So using Epoch::MAX as a "default" value for None is the best.

I don't think AccountHash has much meaning for a default.  It's more because of the `hash()` API under StoredAccountMeta.

Reply:
> Yeah, so for now rent epoch is used, so we should not interpret some magic value to mean 'none'. Same thing for AccountHash.

This is just reason one.  Another reason is probably more important:

> Epoch::MAX means the account is rent-exempt, and the majority of the accounts satisfy this. So using Epoch::MAX as a "default" value for None is the best.

So if the majority of the accounts have their rent-epoch Epoch::MAX (including non-zero-lamport-accounts), then we can use one bit to avoid saving the extra rent-epoch.

Reply:
Offline discussed with @brooksprumo.  So we will use None to express Epoch::MAX to save extra bytes but move the logic from write_account() to write_accounts() to make write_account() take Option<> instead.

Reply:
Done.